internal/services/applications/application_app_role_resource_msgraph.go

```diff
@@ -0,0 +1,198 @@
+package applications
+
+import (
+	"context"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	grapherrors "github.com/manicminer/hamilton/errors"
+	"github.com/manicminer/hamilton/models"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	"github.com/hashicorp/terraform-provider-azuread/internal/helpers/msgraph"
+	"github.com/hashicorp/terraform-provider-azuread/internal/services/applications/parse"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+	"github.com/hashicorp/terraform-provider-azuread/internal/utils"
+)
+
+func applicationAppRoleResourceCreateUpdateMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).Applications.MsClient
+
+	objectId := d.Get("application_object_id").(string)
+
+	var roleId string
+	if v, ok := d.GetOk("role_id"); ok {
+		roleId = v.(string)
+	} else {
+		rid, err := uuid.GenerateUUID()
+		if err != nil {
+			return tf.ErrorDiagF(err, "Generating App Role for application with object ID %q", objectId)
+		}
+		roleId = rid
+	}
+
+	allowedMemberTypesRaw := d.Get("allowed_member_types").(*schema.Set).List()
+	allowedMemberTypes := make([]string, 0, len(allowedMemberTypesRaw))
+	for _, a := range allowedMemberTypesRaw {
+		allowedMemberTypes = append(allowedMemberTypes, a.(string))
+	}
+
+	role := models.AppRole{
```

Comment:
Thanks this is a great tip and I refactored to use the `msgraph` package for both clients and models which is much nicer to import and use.

At the moment this conflicts with the `helpers/msgraph` package but this will itself be refactored in v2.0 once the aadgraph stuff is out.